src/qutip_cupy/dense.py

```diff
@@ -93,6 +93,211 @@ def adjoint(self):
+            return NotImplemented
+        return CuPyDense._raw_cupy_constructor(cp.add(left._cp, right._cp))
+
+    def __matmul__(left, right):
+        if not isinstance(left, CuPyDense) or not isinstance(right, CuPyDense):
+            return NotImplemented
+        return CuPyDense._raw_cupy_constructor(left._cp @ right._cp)
+
+    def __mul__(left, right):
+        dense, number = (left, right) if isinstance(left, CuPyDense) else (right, left)
+        if not isinstance(number, numbers.Number):
+            return NotImplemented
+        return CuPyDense._raw_cupy_constructor(dense._cp * complex(number))
+
+    def __imul__(self, other):
+
+        self._cp.__imul__(other)
+        return self
+
+    def __truediv__(left, right):
+        dense, number = (left, right) if isinstance(left, CuPyDense) else (right, left)
+        if not isinstance(number, numbers.Number):
+            return NotImplemented
+        return CuPyDense._raw_cupy_constructor(dense._cp.__truediv__(number))
+
+    def __itruediv__(self, other):
+        if not isinstance(other, numbers.Number):
+            return NotImplemented
+        self._cp.__itruediv__(other)
+        return self
+
+    def __neg__(self):
+
+        return CuPyDense._raw_cupy_constructor(self._cp.__neg__())
+
+    def __sub__(left, right):
+        if not isinstance(left, CuPyDense) or not isinstance(right, CuPyDense):
+            return NotImplemented
+        return CuPyDense._raw_cupy_constructor(left._cp - right._cp)
+
+    # def __dealloc__(self):
+    #     if self._deallocate and self.data != NULL:
+    #         PyDataMem_FREE(self.data)
+
+
+# @TOCHECK  emti and empty_like are not dispatched in qutip
+# if we do not use them internally here we should remove them
+def empty(rows, cols, fortran):
+    """
+    Return a new Dense type of the given shape, with the data allocated but
+    uninitialised.
+    """
+    order = "F" if fortran else "C"
+    cparr = cp.empty(shape=(rows, cols), dtype=cp.complex128, order=order)
+    return CuPyDense._raw_cupy_constructor(cparr)
+
+
+def empty_like(other, fortran):
+    """
+    Return a new Dense type of the same shape as the given array.
+    """
+    order = "F" if fortran else "C"
+    cparr = cp.empty_like(other, dtype=cp.complex128, order=order)
+    return CuPyDense._raw_cupy_constructor(cparr)
+
+
+def zeros(rows, cols, fortran):
+    """Return the zero matrix with the given shape."""
+    order = "F" if fortran else "C"
+    cparr = cp.zeros(shape=(rows, cols), dtype=cp.complex128, order=order)
+    return CuPyDense._raw_cupy_constructor(cparr)
+
+
+def identity(dimension, scale=1, fortran=True):
+    """
+    Return a square matrix of the specified dimension, with a constant along
+    the diagonal.  By default this will be the identity matrix, but if `scale`
+    is passed, then the result will be `scale` times the identity.
+    """
+    order = "F" if fortran else "C"
+    if scale != 1:
```

Comment:
```suggestion
    if scale == 1:
```
We should also add tests for `scale=1`, scale=`2`, scale=`1+5j`.

Reply:
changed and the tests are being habdled by qutip